File: sdks/javascript/src/registration.ts
import axios from 'axios';
import { KeyPair } from './signing';
import { storeCredentials, storeOAuthToken } from './credentials';
import {
  OAuthProvider,
  getAuthorizationUrl,
  startCallbackServer,
  exchangeCodeForToken,
  openBrowser,
} from './oauth';

export interface RegisterOptions {
  name: string;
  type?: string;
  description?: string;
  version?: string;
  repositoryUrl?: string;
  documentationUrl?: string;
  apiKey?: string;
  oauthProvider?: OAuthProvider;
  redirectUrl?: string;
}

export interface AgentRegistration {
  id: string;
  name: string;
  apiKey: string;
  publicKey: string;
  privateKey?: string; // Only returned on initial registration
  trustScore?: number;
  status?: string;
}

/**
 * Secure is an alias for registerAgent
 * One-line agent registration
 */
export async function secure(
  apiUrl: string,
  options: RegisterOptions
): Promise<AgentRegistration> {
  return registerAgent(apiUrl, options);
}

/**
 * Register a new agent with the AIM backend
 * 
 * IMPORTANT: Agent registration requires a valid API key from a registered user.
 * 
 * To get an API key:
 * 1. Register/login to the AIM dashboard
 * 2. Navigate to Dashboard → API Keys
 * 3. Generate a new API key
 * 4. Pass the API key in the options.apiKey field
 * 
 * The backend generates Ed25519 keypairs automatically and returns credentials.
 * 
 * @param apiUrl - The AIM backend URL (e.g., http://localhost:8080)
 * @param options - Registration options including name, type, and API key
 * @returns Agent registration details including credentials (private key only returned once!)
 */
export async function registerAgent(
  apiUrl: string,
  options: RegisterOptions & { apiKey: string }
): Promise<AgentRegistration> {
  const { name, type = 'ai_agent', apiKey } = options;

  if (!apiKey) {
    throw new Error(
      'API key is required for agent registration. ' +
      'Get your API key from the AIM dashboard: Dashboard → API Keys'
    );
  }

  // Prepare registration payload
  const payload: Record<string, any> = {
    name,
    display_name: name,
    description: options.description || `${name} - AI Agent`,
    agent_type: type,
    version: options.version || '1.0.0',
    repository_url: options.repositoryUrl || '',
    documentation_url: options.documentationUrl || '',
  };

  // Send registration request with API key
  const response = await axios.post(`${apiUrl}/api/v1/public/agents/register`, payload, {
    headers: {
      'Content-Type': 'application/json',
      'X-AIM-API-Key': apiKey,
    },
  });

  const result = response.data;

  // Extract credentials from response
  const agentRegistration: AgentRegistration = {
    id: result.agent_id || result.agentID,
    name: result.name,
    apiKey: apiKey, // Use the same API key for subsequent operations
    publicKey: result.public_key || result.publicKey,
    privateKey: result.private_key || result.privateKey, // ⚠️ Only returned once!
  };

  // Store credentials securely in system keyring
  try {
    await storeCredentials({
      agentId: agentRegistration.id,
      apiKey: agentRegistration.apiKey,
      privateKey: Buffer.from(agentRegistration.privateKey || '', 'base64'),
    });
  } catch (error) {
    console.warn('Warning: Could not store credentials in system keyring:', error);
    console.warn('You will need to manage credentials manually.');
  }

  return agentRegistration;
}

/**
 * SecureWithOAuth is an alias for registerAgentWithOAuth
 */
export async function secureWithOAuth(
  apiUrl: string,
  options: RegisterOptions
): Promise<AgentRegistration> {
  return registerAgentWithOAuth(apiUrl, options);
}

/**
 * Register an agent using OAuth authentication
 * Opens browser for OAuth consent and completes registration
 */
export async function registerAgentWithOAuth(
  apiUrl: string,
  options: RegisterOptions
): Promise<AgentRegistration> {
  const { name, type = 'ai_agent', oauthProvider, redirectUrl = 'http://localhost:8080/callback' } = options;

  if (!oauthProvider) {
    throw new Error('OAuth provider is required for OAuth registration');
  }

  // Get OAuth configuration from environment
  const clientId = process.env[`${oauthProvider.toUpperCase()}_CLIENT_ID`];
  const clientSecret = process.env[`${oauthProvider.toUpperCase()}_CLIENT_SECRET`];

  if (!clientId || !clientSecret) {
    throw new Error(`Missing OAuth credentials for ${oauthProvider}. Set environment variables.`);
  }

  // Generate authorization URL
  const authUrl = getAuthorizationUrl(oauthProvider, clientId, redirectUrl);
  console.log(`🔐 Opening browser for authorization...`);

  // Open browser
  await openBrowser(authUrl);

  // Wait for OAuth callback
  const code = await startCallbackServer(8080);

  // Exchange code for access token
  const token = await exchangeCodeForToken(
    oauthProvider,
    code,
    clientId,
    clientSecret,
    redirectUrl
  );

  // Generate Ed25519 keypair
  const keyPair = KeyPair.generate();

  // Prepare registration payload with OAuth token
  const payload: Record<string, any> = {
    name,
    type,
    public_key: keyPair.publicKeyBase64(),
    oauth_provider: oauthProvider,
    oauth_token: token.accessToken,
  };

  // Sign the payload
  const signature = keyPair.signPayload(payload);
  payload.signature = signature;

  // Send registration request with OAuth token
  const response = await axios.post(`${apiUrl}/api/v1/agents/register`, payload, {
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${token.accessToken}`,
    },
  });

  const result: AgentRegistration = response.data;

  // Store credentials and OAuth token
  await storeCredentials({
    agentId: result.id,
    apiKey: result.apiKey,
    privateKey: keyPair.privateKey,
  });

  await storeOAuthToken(token.accessToken);

  return result;
}

File: sdks/go/registration.go
package aimsdk

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
)

// RegisterOptions configures agent registration
type RegisterOptions struct {
	Name          string        // Agent name (required)
	Type          string        // Agent type: "ai_agent", "mcp_server", etc.
	OAuthProvider OAuthProvider // Optional: OAuth provider for authentication
	RedirectURL   string        // OAuth redirect URL (default: http://localhost:8080/callback)
}

// AgentRegistration holds the registration response
type AgentRegistration struct {
	ID        string `json:"id"`
	Name      string `json:"name"`
	APIKey    string `json:"api_key"`
	PublicKey string `json:"public_key"`
}

// Secure is an alias for RegisterAgent
// Registers an agent with one line of code
func (c *Client) Secure(ctx context.Context, opts RegisterOptions) (*AgentRegistration, error) {
	return c.RegisterAgent(ctx, opts)
}

// RegisterAgent registers a new agent with the AIM backend
// This method generates Ed25519 keypair, signs the request, and stores credentials securely
func (c *Client) RegisterAgent(ctx context.Context, opts RegisterOptions) (*AgentRegistration, error) {
	// Set defaults
	if opts.Type == "" {
		opts.Type = "ai_agent"
	}
	if opts.RedirectURL == "" {
		opts.RedirectURL = "http://localhost:8080/callback"
	}

	// Generate Ed25519 keypair for agent identity
	keyPair, err := GenerateKeyPair()
	if err != nil {
		return nil, fmt.Errorf("keypair generation failed: %w", err)
	}

	// Prepare registration payload
	payload := map[string]interface{}{
		"name":       opts.Name,
		"type":       opts.Type,
		"public_key": keyPair.PublicKeyBase64(),
	}

	// Sign the payload for cryptographic verification
	signature, err := keyPair.SignPayload(payload)
	if err != nil {
		return nil, fmt.Errorf("signing failed: %w", err)
	}
	payload["signature"] = signature

	// Send registration request
	var result AgentRegistration
	if err := c.post(ctx, "/api/v1/agents/register", payload, &result); err != nil {
		return nil, fmt.Errorf("registration request failed: %w", err)
	}

	// Store credentials securely in system keyring
	creds := &Credentials{
		AgentID:    result.ID,
		APIKey:     result.APIKey,
		PrivateKey: keyPair.PrivateKey,
	}

	if err := StoreCredentials(creds); err != nil {
		return nil, fmt.Errorf("credential storage failed: %w", err)
	}

	// Update client configuration with new credentials
	c.config.AgentID = result.ID
	c.config.APIKey = result.APIKey

	return &result, nil
}

// SecureWithOAuth is an alias for RegisterAgentWithOAuth
func (c *Client) SecureWithOAuth(ctx context.Context, opts RegisterOptions) (*AgentRegistration, error) {
	return c.RegisterAgentWithOAuth(ctx, opts)
}

// RegisterAgentWithOAuth registers an agent using OAuth authentication
// This method opens a browser for OAuth consent and completes the registration
func (c *Client) RegisterAgentWithOAuth(ctx context.Context, opts RegisterOptions) (*AgentRegistration, error) {
	// Get OAuth configuration
	oauthConfig, err := GetOAuthConfig(opts.OAuthProvider, opts.RedirectURL)
	if err != nil {
		return nil, fmt.Errorf("OAuth config error: %w", err)
	}

	// Start OAuth flow
	authURL, state := StartOAuthFlow(oauthConfig)
	fmt.Printf("🔐 Please visit this URL to authorize:\n%s\n", authURL)

	// Open browser automatically
	OpenBrowser(authURL)

	// Extract port from redirect URL
	port := 8080 // Default

	// Start callback server to receive OAuth code
	code, err := StartCallbackServer(ctx, port, state)
	if err != nil {
		return nil, fmt.Errorf("OAuth callback failed: %w", err)
	}

	// Exchange code for access token
	token, err := ExchangeCodeForToken(ctx, oauthConfig, code)
	if err != nil {
		return nil, fmt.Errorf("token exchange failed: %w", err)
	}

	// Generate Ed25519 keypair
	keyPair, err := GenerateKeyPair()
	if err != nil {
		return nil, fmt.Errorf("keypair generation failed: %w", err)
	}

	// Prepare registration payload with OAuth token
	payload := map[string]interface{}{
		"name":           opts.Name,
		"type":           opts.Type,
		"public_key":     keyPair.PublicKeyBase64(),
		"oauth_provider": string(opts.OAuthProvider),
		"oauth_token":    token.AccessToken,
	}

	// Sign the payload
	signature, err := keyPair.SignPayload(payload)
	if err != nil {
		return nil, fmt.Errorf("signing failed: %w", err)
	}
	payload["signature"] = signature

	// Send registration request with OAuth token
	var result AgentRegistration
	if err := c.postWithAuth(ctx, "/api/v1/agents/register", payload, token.AccessToken, &result); err != nil {
		return nil, fmt.Errorf("registration request failed: %w", err)
	}

	// Store credentials and OAuth token
	creds := &Credentials{
		AgentID:    result.ID,
		APIKey:     result.APIKey,
		PrivateKey: keyPair.PrivateKey,
	}

	if err := StoreCredentials(creds); err != nil {
		return nil, fmt.Errorf("credential storage failed: %w", err)
	}

	// Store OAuth token separately
	if err := StoreOAuthToken(token.AccessToken); err != nil {
		// Log warning but don't fail
		fmt.Printf("Warning: failed to store OAuth token: %v\n", err)
	}

	// Update client configuration
	c.config.AgentID = result.ID
	c.config.APIKey = result.APIKey

	return &result, nil
}

// post sends a POST request to the API
func (c *Client) post(ctx context.Context, endpoint string, data interface{}, result interface{}) error {
	jsonData, err := json.Marshal(data)
	if err != nil {
		return fmt.Errorf("failed to marshal request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, "POST", c.config.APIURL+endpoint, bytes.NewBuffer(jsonData))
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	if c.config.APIKey != "" {
		req.Header.Set("Authorization", "Bearer "+c.config.APIKey)
	}

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated {
		return fmt.Errorf("API error: status %d", resp.StatusCode)
	}

	if result != nil {
		if err := json.NewDecoder(resp.Body).Decode(result); err != nil {
			return fmt.Errorf("failed to decode response: %w", err)
		}
	}

	return nil
}

// postWithAuth sends a POST request with OAuth bearer token
func (c *Client) postWithAuth(ctx context.Context, endpoint string, data interface{}, token string, result interface{}) error {
	jsonData, err := json.Marshal(data)
	if err != nil {
		return fmt.Errorf("failed to marshal request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, "POST", c.config.APIURL+endpoint, bytes.NewBuffer(jsonData))
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+token)

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated {
		return fmt.Errorf("API error: status %d", resp.StatusCode)
	}

	if result != nil {
		if err := json.NewDecoder(resp.Body).Decode(result); err != nil {
			return fmt.Errorf("failed to decode response: %w", err)
		}
	}

	return nil
}
